### ISSUE TITLE
can't allow a single needless_collect

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#7202]
I tried this code:

```rust
fn main() {
    let v = vec![1, 2, 3, 2, 4];
    #[allow(clippy::needless_collect)]
    let v: Vec<i32> = v.into_iter().map(|x| x * 2).collect();
}
```
This code should not warn but it does.
````
warning: unused variable: `v`
 --> src/main.rs:4:9
  |
4 |     let v: Vec<i32> = v.into_iter().map(|x| x * 2).collect();
  |         ^ help: if this is intentional, prefix it with an underscore: `_v`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: avoid using `collect()` when not needed
 --> src/main.rs:4:52
  |
4 |     let v: Vec<i32> = v.into_iter().map(|x| x * 2).collect();
  |                       -------------                ^^^^^^^
  |                       |
  |                       the iterator could be used here instead
  |
  = note: `#[warn(clippy::needless_collect)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
help: use the original Iterator instead of collecting it and then producing a new one
  |
4 |     v.into_iter().map(|x| x * 2).map(|x| x * 2).collect();
  |    --                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 2 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.68s
````

A global `#![allow(clippy::needless_collect)]` works though.. :thinking: 

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=67d84a3a1892e64fc1d2abc6ffc43101)

### Meta
`clippy 0.1.54 (881c1ac 2021-05-08)`
